src/serac/physics/materials/hyperelastic_material.hpp

```diff
@@ -0,0 +1,308 @@
+// other Serac Project Developers. See the top-level LICENSE file for
+// details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+/**
+ * @file hyperelastic_material.hpp
+ *
+ * @brief The hyperelastic material models for the solid module
+ */
+
+#pragma once
+
+#include "mfem.hpp"
+
+#include "serac/numerics/array_4D.hpp"
+
+namespace serac {
+
+/**
+ * @brief Abstract interface class for a generic hyperelastic material
+ *
+ */
+class HyperelasticMaterial {
+public:
+  /**
+   * @brief Construct a new Hyperelastic Material object
+   *
+   */
+  HyperelasticMaterial() : parent_to_reference_transformation_(nullptr) {}
+
+  /**
+   * @brief Destroy the Hyperelastic Material object
+   *
+   */
+  virtual ~HyperelasticMaterial() = default;
+
+  /// A parent-element to reference-element transformation that can be used to
+  /// evaluate mfem::Coefficient%s.
+  /** @note It is assumed that parent_to_reference_transformation_.SetIntPoint() is already called for the
+      point of interest. */
+
+  /**
+   * @brief Set the reference-to-target transformation. This is required to use coefficient parameters.
+   *
+   * @param[in] parent_to_reference_transformation_ The reference-to-target (stress-free) transformation
+   */
+  void setTransformation(mfem::ElementTransformation& Ttr) { parent_to_reference_transformation_ = &Ttr; }
+
+  /**
+   * @brief Evaluate the strain energy density function, W = W(F).
+   *
+   * @param[in] du_dX the displacement gradient
+   * @return double Strain energy density
+   */
+  virtual double evalStrainEnergy(const mfem::DenseMatrix& du_dX) const = 0;
+
+  /**
+   * @brief Evaluate the Cauchy stress sigma = sigma(F).
+   *
+   * @param[in] du_dX the displacement gradient
+   * @param[out] sigma The evaluated Cauchy stress
+   */
+  virtual void evalStress(const mfem::DenseMatrix& du_dX, mfem::DenseMatrix& sigma) const = 0;
+
+  /**
+   * @brief Evaluate the derivative of the 1st Piola-Kirchhoff stress tensor in Voigt notation
+   * and assemble its contribution to the local gradient matrix 'A'.
+
+   * @param[in] du_dX the displacement gradient
+   * @param[out] C Tangent moduli 4D Array in spatial form (C^e_ijkl=(d tau_ij)/(d F_km) * F_lm = J * sigma_ij delta_kl
+   + J * (d sigma_ij)/(d F_km) F_lm )
+   */
+  virtual void evalTangentStiffness(const mfem::DenseMatrix& du_dX, serac::mfem_ext::Array4D<double>& C) const = 0;
+
+  /**
+   * @brief Calculate the deformation gradient from the displacement gradient (F = H + I)
+   *
+   * @param[in] du_dX the displacement gradient (du_dX)
+   * @param[out] F the deformation gradient (dx_dX)
+   */
+  static void calcDeformationGradient(const mfem::DenseMatrix& du_dX, mfem::DenseMatrix& F);
+
+  /**
+   * @brief Calculate the linearized strain tensor (epsilon = 1/2 * (du_dX + du_dX^T))
+   *
+   * @param[in] du_dX the displacement gradient (du_dX)
+   * @param[out] epsilon the linearized strain tensor epsilon = 1/2 * (du_dX + du_dX^T)
```

Comment:
these are useful conversions, but aren't really specific to Hyperelasticity. To your point earlier, I wonder where we should put things like this. Maybe `solid_mechanics::calcDeformationGradient`?

Reply:
Yeah, I needed to put them here instead of the solid class to avoid a circular dependency. Any ideas @joshessman-llnl @white238 ?

Reply:
Another directory for useful calculations/conversions? physics_utilities?

Reply:
Then the problem is how do we connect them with a specific physics. In other words, we have the `serac::Solid` class, so we would use the `serac::solid` namespace. Alternatively we can lump all the physics together and use `serac::physics_utils`.

Reply:
(I tried to add them as static members to the `Solid` class, but that's when I noticed the circular dependency.)

Reply:
I ended up putting them in a `solid_util` namespace. Suggestions welcome!